### PR TITLE
add stable structures tests, make global init and post upgrade state …

### PR DESCRIPTION
**Pull request by lastmjs**
…better

### CODE REVIEW

File: tests/end_to_end/candid_rpc/class_syntax/stable_structures/src/canister1/index.ts [hunk 1]
```diff
@@ -1,42 +1,251 @@
...
+let stableMap2 = StableBTreeMap<number, bigint>(2);
+let stableMap3 = StableBTreeMap<Reaction, bigint>(3);
+let stableMap4 = StableBTreeMap<User, number>(4);
 
-export default class extends StableMap0Methods {
+export default class {
     @query([], IDL.Bool)
     getRedeployed(): boolean {
-        return globalThis._azlePostUpgradeCalled;
+        return globalThis._azlePostUpgradeCalled; // TODO this should be set in the Rust init and post upgrade so that they are always set
```
Comment (bdemann) [inline tests/end_to_end/candid_rpc/class_syntax/stable_structures/src/canister1/index.ts:14]: get rid of this TODO

File: src/lib/stable/post_upgrade.ts [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+export function postUpgrade(paramIdls: IDL.Type[]): MethodDecorator {
+    return <T>(
+        target: object,
+        propertyKey: string | symbol,
+        descriptor: TypedPropertyDescriptor<T>
+    ): TypedPropertyDescriptor<T> | void => {
+        const originalMethod = (descriptor.value as any).bind(target);
+
+        const methodCallback = (...args: any[]) => {
+            globalThis._azlePostUpgradeCalled = true;
```
Comment (bdemann) [inline src/lib/stable/post_upgrade.ts:14]: can we get rid of this everywhere?

**Review by bdemann — CHANGES_REQUESTED** [review]

File: src/lib/stable/post_upgrade.ts (hunk 2 above)
Comment (lastmjs) [reply src/lib/stable/post_upgrade.ts:14]: Fixed everywhere in subsequent PRs

File: tests/end_to_end/candid_rpc/class_syntax/stable_structures/src/canister1/index.ts (hunk 1 above)
Comment (lastmjs) [reply tests/end_to_end/candid_rpc/class_syntax/stable_structures/src/canister1/index.ts:14]: This has been removed in a subsequent PR

**Pull request merged**